File: nordgen-go/internal/wireguard/wireguard.go
package wireguard

import (
	"encoding/base64"
	"fmt"
	"net/netip"
	"strconv"
	"strings"
)

const (
	endpointPort              = 51820
	wireGuardKeyEncodedLength = 44
)

func ValidateKey(value string) error {
	if len(value) != wireGuardKeyEncodedLength || value[wireGuardKeyEncodedLength-1] != '=' {
		return fmt.Errorf("key must be a base64-encoded 32-byte value")
	}

	for index := 0; index < wireGuardKeyEncodedLength-1; index++ {
		character := value[index]
		if (character >= 'A' && character <= 'Z') ||
			(character >= 'a' && character <= 'z') ||
			(character >= '0' && character <= '9') ||
			character == '+' || character == '/' {
			continue
		}
		return fmt.Errorf("key must be a base64-encoded 32-byte value")
	}

	decoded, err := base64.StdEncoding.DecodeString(value)
	if err != nil || len(decoded) != 32 {
		return fmt.Errorf("key must be a base64-encoded 32-byte value")
	}
	return nil
}

func ValidateEndpoint(value string) error {
	value = strings.TrimSpace(value)
	if value == "" {
		return fmt.Errorf("endpoint is empty")
	}
	if strings.ContainsAny(value, "\r\n\t ") {
		return fmt.Errorf("endpoint contains whitespace")
	}
	if _, err := netip.ParseAddr(value); err == nil {
		return nil
	}
	if len(value) > 253 || strings.HasSuffix(value, ".") {
		return fmt.Errorf("endpoint hostname is invalid")
	}
	labels := strings.Split(value, ".")
	if len(labels) < 2 {
		return fmt.Errorf("endpoint hostname is invalid")
	}
	for _, label := range labels {
		if len(label) == 0 || len(label) > 63 || label[0] == '-' || label[len(label)-1] == '-' {
			return fmt.Errorf("endpoint hostname is invalid")
		}
		for index := 0; index < len(label); index++ {
			character := label[index]
			if (character >= 'a' && character <= 'z') ||
				(character >= 'A' && character <= 'Z') ||
				(character >= '0' && character <= '9') ||
				character == '-' {
				continue
			}
			return fmt.Errorf("endpoint hostname is invalid")
		}
	}
	return nil
}

func BuildConfig(privateKey, publicKey, endpoint, dns string, keepalive int) ([]byte, error) {
	if err := ValidateKey(privateKey); err != nil {
		return nil, fmt.Errorf("invalid private key: %w", err)
	}
	if err := ValidateKey(publicKey); err != nil {
		return nil, fmt.Errorf("invalid public key: %w", err)
	}
	endpoint = strings.TrimSpace(endpoint)
	if err := ValidateEndpoint(endpoint); err != nil {
		return nil, fmt.Errorf("invalid endpoint: %w", err)
	}
	dnsAddr, err := netip.ParseAddr(strings.TrimSpace(dns))
	if err != nil {
		return nil, fmt.Errorf("invalid DNS address")
	}
	if keepalive < 0 || keepalive > 65535 {
		return nil, fmt.Errorf("keepalive must be between 0 and 65535 seconds")
	}

	host := endpoint
	if addr, parseErr := netip.ParseAddr(endpoint); parseErr == nil {
		host = addr.String()
		if addr.Is6() {
			host = "[" + host + "]"
		}
	}

	var builder strings.Builder
	builder.Grow(256)
	builder.WriteString("[Interface]\nPrivateKey = ")
	builder.WriteString(privateKey)
	builder.WriteString("\nAddress = 10.5.0.2/16\nDNS = ")
	builder.WriteString(dnsAddr.String())
	builder.WriteString("\n\n[Peer]\nPublicKey = ")
	builder.WriteString(publicKey)
	builder.WriteString("\nAllowedIPs = 0.0.0.0/0, ::/0\nEndpoint = ")
	builder.WriteString(host)
	builder.WriteByte(':')
	builder.WriteString(strconv.Itoa(endpointPort))
	builder.WriteString("\nPersistentKeepalive = ")
	builder.WriteString(strconv.Itoa(keepalive))
	builder.WriteByte('\n')

	return []byte(builder.String()), nil
}

File: nordgen/src/nord_config_generator/wireguard.py
import base64
import binascii
from ipaddress import ip_address

ENDPOINT_PORT = 51820


class WireGuardValueError(ValueError):
    pass


def validate_key(value: str) -> None:
    try:
        decoded = base64.b64decode(value, validate=True)
    except (binascii.Error, ValueError) as error:
        raise WireGuardValueError("key must be a base64-encoded 32-byte value") from error
    if len(decoded) != 32:
        raise WireGuardValueError("key must be a base64-encoded 32-byte value")


def validate_endpoint(value: str) -> None:
    endpoint = value.strip()
    if not endpoint:
        raise WireGuardValueError("endpoint is empty")
    if any(character.isspace() for character in endpoint):
        raise WireGuardValueError("endpoint contains whitespace")
    try:
        ip_address(endpoint)
        return
    except ValueError:
        pass
    if len(endpoint) > 253 or endpoint.endswith("."):
        raise WireGuardValueError("endpoint hostname is invalid")
    labels = endpoint.split(".")
    if len(labels) < 2:
        raise WireGuardValueError("endpoint hostname is invalid")
    for label in labels:
        if not label or len(label) > 63 or label.startswith("-") or label.endswith("-"):
            raise WireGuardValueError("endpoint hostname is invalid")
        if not all(
            character.isascii() and (character.isalnum() or character == "-") for character in label
        ):
            raise WireGuardValueError("endpoint hostname is invalid")


def build_config(
    private_key: str,
    public_key: str,
    endpoint: str,
    dns: str,
    keepalive: int,
) -> bytes:
    try:
        validate_key(private_key)
    except WireGuardValueError as error:
        raise WireGuardValueError(f"invalid private key: {error}") from error
    try:
        validate_key(public_key)
    except WireGuardValueError as error:
        raise WireGuardValueError(f"invalid public key: {error}") from error

    normalized_endpoint = endpoint.strip()
    try:
        validate_endpoint(normalized_endpoint)
    except WireGuardValueError as error:
        raise WireGuardValueError(f"invalid endpoint: {error}") from error

    try:
        dns_address = ip_address(dns.strip())
    except ValueError as error:
        raise WireGuardValueError("invalid DNS address") from error
    if keepalive < 0 or keepalive > 65535:
        raise WireGuardValueError("keepalive must be between 0 and 65535 seconds")

    host = normalized_endpoint
    try:
        endpoint_address = ip_address(normalized_endpoint)
    except ValueError:
        endpoint_address = None
    if endpoint_address is not None:
        host = endpoint_address.compressed
        if endpoint_address.version == 6:
            host = f"[{host}]"

    content = (
        f"[Interface]\nPrivateKey = {private_key}\n"
        f"Address = 10.5.0.2/16\nDNS = {dns_address.compressed}\n\n"
        f"[Peer]\nPublicKey = {public_key}\n"
        f"AllowedIPs = 0.0.0.0/0, ::/0\n"
        f"Endpoint = {host}:{ENDPOINT_PORT}\n"
        f"PersistentKeepalive = {keepalive}\n"
    )
    return content.encode()
